### PR TITLE
feat: add protobuf codec and migrate tail based sampling to protobuf

### PR DESCRIPTION
**Pull request by kruskall**
## Motivation/summary

Add protobuf codec
Update tail-based sampling  to use protobufcodec

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## Related issues

Blocked by https://github.com/elastic/apm-server/pull/10981
Related to https://github.com/elastic/apm-data/issues/52

### CODE REVIEW

File: x-pack/apm-server/sampling/eventstorage/protobufcodec.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+package eventstorage
+
+import (
+	"google.golang.org/protobuf/proto"
+
+	"github.com/elastic/apm-data/model/modelpb"
+)
+
+// JSONCodec is an implementation of Codec, using JSON encoding.
```
Comment (axw) [inline x-pack/apm-server/sampling/eventstorage/protobufcodec.go:13]:
```suggestion
// ProtobufCodec is an implementation of Codec, using protobuf encoding.
```

File: x-pack/apm-server/sampling/eventstorage/protobufcodec.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+type ProtobufCodec struct{}
+
+// DecodeEvent decodes data as JSON into event.
```
Comment (axw) [inline x-pack/apm-server/sampling/eventstorage/protobufcodec.go:16]:
```suggestion
// DecodeEvent decodes data as protobuf into event.
```

File: x-pack/apm-server/sampling/eventstorage/protobufcodec.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+func (ProtobufCodec) DecodeEvent(data []byte, event *modelpb.APMEvent) error {
+	return proto.Unmarshal(data, event)
+}
+
+// EncodeEvent encodes event as JSON.
```
Comment (axw) [inline x-pack/apm-server/sampling/eventstorage/protobufcodec.go:21]:
```suggestion
// EncodeEvent encodes event as protobuf.
```

**Review by axw — APPROVED** [review]
:tada:

File: x-pack/apm-server/sampling/eventstorage/protobufcodec.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kruskall) [reply x-pack/apm-server/sampling/eventstorage/protobufcodec.go]: Fixed :+1:

Comment (kruskall) [reply x-pack/apm-server/sampling/eventstorage/protobufcodec.go]: :facepalm:

**Review by dmathieu — APPROVED** [review]

**Comment by marclop** [comment]
Late to the party but look at those benchmarks! 🎉 

```
pkg: github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage
                                            │ build/main/bench.out │              bench.out               │
                                            │        sec/op        │    sec/op      vs base               │
ShardedWriteTransactionUncontended-12               156.20n ±  11%   97.07n ±  15%  -37.86% (p=0.002 n=6)
ShardedWriteTransactionContended-12                 2138.0n ±  43%   859.5n ± 112%  -59.80% (p=0.026 n=6)
WriteTransaction/json_codec-12                       2.971µ ±  15%   1.548µ ±  41%  -47.90% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/1_events-12             6.505µ ±  10%   5.379µ ±  13%  -17.32% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/100_events-12           439.8µ ±   5%   338.9µ ±  10%  -22.94% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/199_events-12           870.4µ ±   7%   704.5µ ±  14%  -19.05% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/399_events-12           1.610m ±   4%   1.230m ±   5%  -23.62% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/1000_events-12          3.850m ±   7%   2.981m ±   4%  -22.58% (p=0.002 n=6)
IsTraceSampled/unsampled-12                          79.92n ±   1%   78.17n ±   1%   -2.19% (p=0.002 n=6)
geomean                                              16.89µ          14.50µ         -14.18%

                                            │ build/main/bench.out │               bench.out                │
                                            │         B/op         │     B/op       vs base                 │
ShardedWriteTransactionUncontended-12                  505.0 ±  3%     446.5 ±  4%  -11.58% (p=0.002 n=6)
ShardedWriteTransactionContended-12                  1.430Ki ± 39%   1.040Ki ± 23%  -27.25% (p=0.026 n=6)
WriteTransaction/json_codec-12                         309.0 ±  0%     285.0 ±  0%   -7.77% (p=0.002 n=6)
WriteTransaction/json_codec_big_tx-12                 1453.5 ±  0%     997.0 ±  0%  -31.41% (p=0.002 n=6)
ReadEvents/json_codec/1_events-12                    1.882Ki ±  0%   1.849Ki ±  0%   -1.74% (p=0.002 n=6)
ReadEvents/json_codec/10_events-12                   12.01Ki ±  1%   11.67Ki ±  0%   -2.87% (p=0.002 n=6)
ReadEvents/json_codec/100_events-12                  112.9Ki ±  0%   109.5Ki ±  0%   -3.06% (p=0.002 n=6)
ReadEvents/json_codec/199_events-12                  183.6Ki ±  0%   180.9Ki ±  1%   -1.51% (p=0.002 n=6)
ReadEvents/json_codec/399_events-12                  308.7Ki ±  0%   304.0Ki ±  1%   -1.52% (p=0.002 n=6)
ReadEvents/json_codec/1000_events-12                 659.7Ki ±  0%   655.7Ki ±  1%   -0.60% (p=0.009 n=6)
ReadEvents/json_codec_big_tx/1_events-12             6.034Ki ±  0%   4.919Ki ±  0%  -18.48% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/10_events-12            53.47Ki ±  0%   42.33Ki ±  0%  -20.84% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/100_events-12           528.2Ki ±  0%   416.5Ki ±  0%  -21.14% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/199_events-12           825.1Ki ±  0%   714.6Ki ±  0%  -13.40% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/399_events-12           1.372Mi ±  0%   1.265Mi ±  0%   -7.75% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/1000_events-12          3.051Mi ±  0%   2.949Mi ±  0%   -3.33% (p=0.002 n=6)
geomean                                              11.96Ki         11.35Ki         -5.11%
¹ all samples are equal

                                            │ build/main/bench.out │              bench.out              │
                                            │      allocs/op       │  allocs/op   vs base                │
ReadEvents/json_codec_big_tx/1_events-12                86.00 ± 0%    85.00 ± 0%  -1.16% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/10_events-12               735.0 ± 0%    725.0 ± 0%  -1.36% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/100_events-12             7.215k ± 0%   7.115k ± 0%  -1.39% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/199_events-12             13.85k ± 0%   13.65k ± 0%  -1.44% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/399_events-12             27.05k ± 0%   26.65k ± 0%  -1.48% (p=0.002 n=6)
ReadEvents/json_codec_big_tx/1000_events-12            66.51k ± 0%   65.51k ± 0%  -1.51% (p=0.002 n=6)
geomean                                                 127.7         127.4       -0.23%
¹ all samples are equal
```

**Pull request merged**